### ISSUE TITLE
Improve the usability of cachedQuery

### ISSUE THREAD

**Issue by Steven-Chan** [SkygearIO/features#135]
## Problem

JS SDK API call returns a promise for client app to get result when the promise is resolved.

However, promise can only be resolved once, this pattern does not fit the cached result, i.e. the API call resolve a cached result first, and then remote result.

## Simple and Dumb solution proposed by a few other team member

```js
const resolver = (cached, result) => {
  // do something with result
};
skygear.publicDB.query(query).then(resolver.bind(this, false));
skygear.publicDB.queryCache(query).then(resolver.bind(this, true));
```

### Old function signature
@chpapa also suggested that we can keep the old function signature for:
1. Backward compatibility
1. It is actually better if the cachedCallback won't be called if the real result arrive faster (although unlikely)

Old function signature:
```js
function successCallback(notes, cached = false) {
  if (cached) {
    console.log('The result is a cached one');
  }
  console.log('Result', notes);
}
skygear.publicDB.query(query, successCallback)
.then(successCallback, (error) => {
  console.error(error);
});
```

### Other considerations

Some other considerations have been raised by fellows:
1. Shall we allow user to invalidate a cache query?
    * If yes, it will probably be a 3rd function call like `skygear.publicDB.queryCacheInvalidate(query)`
1. Shall we set default cache, or default no-cache, for `skygear.publicDB.query` to update the cache?
    * No matter what, it would change the function signature to `skygear.publicDB.query(query, cachedCallback, cached=True)`

### Overview

current:

```js
func functionName(params) -> Promise

skygear.functionName(params)
  .then((result) => {
    // on result resolved
  });
```

with Rx:

```js
func functionName(params) -> Observable

skygear.functionName(params)
  .subscribe((result) => {
    // on new result get
    // 
    // this can be triggered more than once
    // when cached result received and fetched result received
  });
```

Pros:

- Rx is a common way to handle async event, like Angular 2 changed its API to return Observable

Cons:

- Migration may be tricky if there are a lot of promise chain
- Rx is more complicated than Promise, like Observable needs to be disposed

### Backward compatibility with Promise

I think it is reasonable for the app to decide EITHER using Promise or Rx.

Default skygear container may use Rx. Developer requires to call a function to make async API functions to return a Promise instead.

Something like:

```
skygear.usePromise()
```

Basically, skygear would call `toPromise` on the `Observable`. Thus only the last value of observable would be resolved.
